peps/pep-0705.rst

```diff
@@ -77,106 +82,297 @@ The problem disappears if we don't have mutator methods in ``Movie``. This could
+``TypedDict`` will gain two new boolean flags: ``other_keys`` and ``readonly``. A new ``typing.ReadOnly`` type qualifier is added.
 
-Notable similarities to ``TypedDict``:
+``other_keys`` flag
+-------------------
 
-* A ``TypedMapping`` protocol can be declared using class-based or alternative syntax.
-* Keys must be strings.
-* By default, all specified keys must be present in a ``TypedMapping`` instance. It is possible to override this by specifying totality, or by using ``NotRequired`` from :pep:`655`.
-* Methods are not allowed in the declaration (though they may be inherited).
+The ``other_keys`` flag to ``TypedDict``, when ``False``, indicates that instances may only contain keys explicitly listed in the type::
 
-Notable differences from ``TypedDict``:
+    class Album(TypedDict, other_keys=False):
+        name: str
+        year: int
 
-* The runtime type of a ``TypedMapping`` object is not constrained to be a ``dict``.
-* No mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated.
-* The ``|`` operator is not supported.
-* A class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases.
-* Subclasses can narrow value types, in the same manner as other protocols.
+    class AlbumExtra(Album, TypedDict):
+        band: str  # Runtime error
 
-As with :pep:`589`, this PEP provides a sketch of how a type checker is expected to support type checking operations involving ``TypedMapping`` and ``TypedDict`` objects, but details are left to implementors. In particular, type compatibility should be based on structural compatibility.
+This flag defaults to ``True``.
 
+Type-checkers may rely on this restriction::
 
-Multiple inheritance and TypedDict
-----------------------------------
+    def album_keys(album: Album) -> Collection[Literal['name'] | Literal['year']]:
+        # Type checkers may permit this, but should error if Album did not specify `other_keys=False`
+        return album.keys()
 
-A type that inherits from a ``TypedMapping`` protocol and from ``TypedDict`` (either directly or indirectly):
+Type-checkers should prevent operations that would violate this restriction::
 
-* is the structural intersection of its parents, or invalid if no such intersection exists
-* instances must be a dict subclass
-* adds mutator methods only for fields it explicitly (re)declares
+    class AlbumExtra(TypedDict, other_keys=False):
+        name: str
+        year: int
+        band: str
+
+    album: AlbumExtra = { "name": "Flood", year: 1990, band: "They Might Be Giants" }
+    album_keys(album)  # Type check error: extra key 'band'
+
+``readonly`` flag
+-----------------
 
-For example::
+The ``readonly`` flag to ``TypedDict``, when ``True``, indicates that no mutator methods (``__setitem__``, ``__delitem``, ``update``, etc.) will be generated::
```

Comment:
```suggestion
The ``readonly`` flag to ``TypedDict``, when ``True``, indicates that no mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated::
```

Apart from the typo, the "generated" wording feels slightly wrong because type checkers may not work internally by "generating" these methods. I'd suggest instead:

```suggestion
The ``readonly`` flag to ``TypedDict``, when ``True``, indicates that no mutator operations (``__setitem__``, ``__delitem__``, ``update``, etc.) will be permitted::
```